views/js/runner/runner.js

```diff
@@ -0,0 +1,476 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2015 (original work) Open Assessment Technologies SA ;
+ */
+/**
+ * @author Jean-Sébastien Conan <jean-sebastien.conan@vesperiagroup.com>
+ * @author Sam <sam@taotesting.com>
+ */
+define([
+    'jquery',
+    'lodash',
+    'i18n',
+    'core/eventifier',
+    'core/promise',
+    'core/logger',
+    'taoTests/runner/providerRegistry'
+], function ($, _, __, eventifier, Promise, logger, providerRegistry){
+    'use strict';
+
+    /**
+     * Builds an instance of the QTI test runner
+     *
+     * @param {String} providerName
+     * @param {Object} config
+     * @param {String|DOMElement|JQuery} config.contentContainer - the dom element that is going to holds the test content (item, rubick, etc)
+     * @param {Array} [config.plugins] - the list of plugin instances to be initialized and bound to the test runner
+     * @returns {runner|_L28.testRunnerFactory.runner}
+     */
+    function testRunnerFactory(providerName, pluginFactories, config){
+
+        /**
+         * @type {Object} The test runner instance
+         */
+        var runner;
+
+        /**
+         * @type {Object} the test definition data
+         */
+        var testData       = {};
+
+        /**
+         * @type {Object} contextual test data (the state of the test)
+         */
+        var testContext    = {};
+
+        /**
+         * @type {Object} the registered plugins
+         */
+        var plugins        = {};
+
+        /**
+         * @type {Object} the test of the runner
+         */
+        var states = {
+            'init':    false,
+            'ready':   false,
+            'render':  false,
+            'finish':  false,
+            'destroy': false
+        };
+
+        /**
+         * The selected test runner provider
+         */
+        var provider  = testRunnerFactory.getProvider(providerName);
+
+        var areaBroker;
+
+        /**
+         * Run a method of the provider (by delegation)
+         *
+         * @param {String} method - the method to run
+         * @param {...} args - rest parameters given to the provider method
+         * @returns {Promise} so provider can do async stuffs
+         */
+        function providerRun(method){
+            var args = [].slice.call(arguments, 1);
+            return new Promise(function(resolve){
+                if(!_.isFunction(provider[method])){
+                   resolve();
+                }
+                resolve(provider[method].apply(runner, args));
```

Comment:
Potential double call of the `resolve` method.
Exception may be thrown if the method is not a function.